src/lib/comms/include/sol-gatt.h

```diff
@@ -0,0 +1,256 @@
+	 */
+	int (*read)(struct sol_gatt_pending *op,
+		    uint16_t offset);
+	/**
+	 * The response for this operation will be returned by calling
+	 * #sol_gatt_pending_reply(), the @a buf parameter will be ignored.
+	 */
+	int (*write)(struct sol_gatt_pending *op,
+		     struct sol_buffer *buf,
+		     uint16_t offset);
+	/**
+	 * This operation doesn't require an reply.
+	 */
+	/*
+	 * FIXME: is there any case that it would make sense to
+	 * wait for a reply?
+	 */
+	int (*flush)(struct sol_gatt_pending *op,
+		     bool sync);
+	void *user_data;
+};
+
+/**
+ * @brief Returns a response to a async operation
+ *
+ * When a operation is performed in a attribute, the response to the operation
+ * is only returned when this function is called.
+ *
+ * @param pending Pending operation to be responded
+ * @param error Error to be returned
+ * @param buf Payload (can be NULL) to be returned
+ */
+int sol_gatt_pending_reply(struct sol_gatt_pending *pending,
+			   enum sol_gatt_error error,
+			   struct sol_buffer *buf);
+
+/*
+ * Helpers to the construction of the local attribute database
+ */
+#define SOL_GATT_SERVICE(service_uuid) {			\
+		.type = SOL_GATT_ATTR_TYPE_SERVICE,		\
+		.uuid = service_uuid,				\
+	}
+
+#define SOL_GATT_CHARACTERISTIC(uuid, flags, read, write,		\
+				flush, user_data) {			\
+		.type = SOL_GATT_ATTR_TYPE_CHARACTERISTIC,		\
+		.uuid = uuid,						\
+		.flags = flags,						\
+		.read = read,						\
+		.write = write,						\
+		.flush = flush,						\
+		.user_data = user_data,					\
+	}
+
+#define SOL_GATT_DESCRIPTOR(uuid, flags, read, write, flush, user_data) { \
+		.type = SOL_GATT_ATTR_TYPE_DESCRIPTOR,			\
+		.uuid = uuid,						\
+		.flags = flags,						\
+		.read = read,						\
+		.write = write,						\
+		.flush = flush,						\
+		.user_data = user_data,					\
+	}
+
+/**
+ * @brief Registers attributes into the GATT database
+ *
+ * @param attrs Array of the attributes to be registered
+ * @param len Length of the array of attributes
+ *
+ * @return 0 on success, -errno on failure.
+ */
+int sol_gatt_register_attributes(struct sol_gatt_attr *attrs, uint16_t len);
+
+/**
+ * @brief Unregisters attributes from the GATT database
+ *
+ * @param attrs Array of the attributes to be unregistered
+ * @param len Length of the array of attributes
+ *
+ * @return 0 on success, -errno on failure.
+ */
+int sol_gatt_unregister_attributes(struct sol_gatt_attr *attrs, uint16_t len);
+
+/**
+ * @brief Reads the value from a attribute
+ *
+ * This only really makes sense for Characteristics and Descriptors,
+ * the Read Long attribute procedure will be performed if it's supported.
+ *
+ * @param conn Connection to a remote device
+ * @param attr Attribute to be read
+ * @param cb Callback to be called when the operation is finished
+ * @param user_data User pointer to be passed to the callback
+ */
+int sol_gatt_read_attr(struct sol_bt_conn *conn, struct sol_gatt_attr *attr,
```

Comment:
do we need a pending handle to cancel it? Or it is automatically gone if the connection dies? We need to write that here.

Case is:
 - ask read to show in some context;
 - context is gone (ie: window closed, widgets and local context is free()ed);
 - you must have a way to stop the read, otherwise it will call you back and you'll access data that is not there anymore.

In this case I believe we must return a handle to cancel, otherwise we'd have to kill the connection to support this case, but maybe the other parts of your app still needs that connection